lib/WUI/nhttp/file_info.cpp

```diff
@@ -23,19 +23,116 @@ FileInfo::FileInfo(const char *filename, bool can_keep_alive, bool after_upload)
     : can_keep_alive(can_keep_alive)
     , after_upload(after_upload) {
     strlcpy(this->filename, filename, sizeof this->filename);
+    /*
+     * Eat the last slash on directories.
+     *
+     * This is mostly aestetic only... otherwise we get double slashs in the
+     * results/URLs. That works, but, meh...
+     */
+    const size_t len = strlen(this->filename);
+    if (len > 0 && this->filename[len - 1] == '/') {
+        this->filename[len - 1] = '\0';
+    }
 }
 
 Step FileInfo::step(std::string_view, bool, uint8_t *output, size_t output_size) {
-    // TODO: Support directories.
-
-    // We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
-    // We'll have to revisit this for directories, of course.
+    /*
+     * We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
+     * Directories send each file as a single packet, but that one also assumes it fits.
+     */
     ConnectionHandling handling = can_keep_alive ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
     struct stat finfo;
-    if (stat(filename, &finfo) == 0) {
-        if (finfo.st_mode & S_IFDIR) {
-            return Step { 0, 0, StatusPage(Status::NotImplemented, can_keep_alive, "Directory listing not implemented") };
-        } else {
+    const size_t last_chunk_len = strlen(LAST_CHUNK);
+
+    if (dir) {
+        /*
+         * We already have a directory listing running. Try to pull another file out of there.
+         */
+        struct dirent *ent;
+        while ((ent = readdir(dir.get()))) {
+            // Check this is GCODE, not some other thing. Skip the others.
+            const char *last_dot = strrchr(ent->d_name, '.');
+            if (last_dot == nullptr) {
+                continue;
+            }
+            if (strcasecmp(last_dot, ".gcode") != 0 && strcasecmp(last_dot, ".gco") != 0) {
+                continue;
+            }
+
+            size_t written = render_chunk(handling, output, output_size, [&](uint8_t *output, size_t output_size) {
+                size_t written = 0;
+                if (need_comma) {
+                    assert(output_size > 0);
```

Comment:
What does the assert do in the FW? I know what it does in unit tests :wink:

Reply:
Similar to abort, if it fails, it kills the thread (which can be observed by the thread getting stuck). Then the backtrace is particularly handy, because it points to the exact location where it happened. Much more usable than bluescreen outputs and dumps, actually.

Reply:
ok